File: WorkOrderbackend/app/Models/WorkOrder.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

class WorkOrder extends Model
{
    protected $fillable = [
        'title',
        'description',
        'requester_id',
        'assigned_engineer_id',
        'department_id',
        'priority_id',
        'status',
        'due_at',
        'assigned_at',
        'started_at',
        'completed_at',
        'cancelled_at',
        'overdue_at',
        'cancelled_by',
        'cancelled_role',
    ];

    protected $casts = [
        'due_at' => 'datetime',
        'assigned_at' => 'datetime',
        'started_at' => 'datetime',
        'completed_at' => 'datetime',
        'cancelled_at' => 'datetime',
        'overdue_at' => 'datetime',
    ];

    public function requester()
    {
        return $this->belongsTo(User::class, 'requester_id');
    }

    public function engineer()
    {
        return $this->belongsTo(User::class, 'assigned_engineer_id');
    }

    public function department()
    {
        return $this->belongsTo(Department::class);
    }

    public function cancelledByUser()
    {
        return $this->belongsTo(User::class, 'cancelled_by');
    }

    public function comments()
    {
        return $this->hasMany(Comment::class)->with('user')->latest('created_at');
    }
}

File: workorderapp/lib/entity/work_order.dart
import 'package:workorderapp/entity/comment.dart';
import 'package:workorderapp/entity/department.dart';
import 'package:workorderapp/entity/user.dart';

class WorkOrder {
  final int id;

  final String title;
  final String description;
  final String status;

  final int requesterId;
  final int? assignedEngineerId;
  final int? departmentId;
  final int? priorityId;

  final DateTime? dueAt;
  final DateTime? assignedAt;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final DateTime? overdueAt;
  final DateTime? cancelledAt;

  final DateTime? createdAt;
  final DateTime? updatedAt;

  final int? cancelledBy;
  final String? cancelledRole;

  final User requester;
  final User? engineer;
  final Department? department;
  final User? cancelledByUser;
  final List<Comment> comments;
  WorkOrder({
    required this.id,
    required this.title,
    required this.description,
    required this.status,
    required this.requesterId,
    required this.requester,
    this.assignedEngineerId,
    this.departmentId,
    this.priorityId,
    this.engineer,
    this.department,
    this.cancelledByUser,
    this.dueAt,
    this.assignedAt,
    this.startedAt,
    this.completedAt,
    this.overdueAt,
    this.cancelledAt,
    this.createdAt,
    this.updatedAt,
    this.cancelledBy,
    this.cancelledRole,
    this.comments = const [],
  });

  factory WorkOrder.fromJson(Map<String, dynamic> json) {
    try {
      int? parseNullableInt(dynamic value) {
        if (value == null) return null;
        if (value is int) return value;
        if (value is String && value.isNotEmpty) return int.tryParse(value);
        return null;
      }

      DateTime? parseNullableDateTime(dynamic value) {
        if (value == null) return null;
        try {
          return DateTime.parse(value.toString()).toLocal();
        } catch (_) {
          return null;
        }
      }

      return WorkOrder(
        id: parseNullableInt(json['id']) ?? 0,
        title: json['title'] ?? '',
        description: json['description'] ?? '',
        status: json['status'] ?? '',
        requesterId: parseNullableInt(json['requester_id']) ?? 0,
        assignedEngineerId: parseNullableInt(json['assigned_engineer_id']),
        departmentId: parseNullableInt(json['department_id']),
        priorityId: parseNullableInt(json['priority_id']),
        requester: json['requester'] != null
            ? User.fromJson(json['requester'])
            : User(
                id: 0,
                name: 'Unknown',
                username: 'unknown@example.com',
                roles: [],
                department: null,
              ),
        engineer: json['engineer'] != null
            ? User.fromJson(json['engineer'])
            : null,
        department: json['department'] != null
            ? Department.fromJson(json['department'])
            : null,
        cancelledByUser: json['cancelled_by_user'] != null
            ? User.fromJson(json['cancelled_by_user'])
            : null,
        comments: (json['comments'] as List? ?? [])
            .map((c) => Comment.fromJson(c))
            .toList(),
        dueAt: parseNullableDateTime(json['due_at']),
        assignedAt: parseNullableDateTime(json['assigned_at']),
        startedAt: parseNullableDateTime(json['started_at']),
        completedAt: parseNullableDateTime(json['completed_at']),
        overdueAt: parseNullableDateTime(json['overdue_at']),
        cancelledAt: parseNullableDateTime(json['cancelled_at']),
        createdAt: parseNullableDateTime(json['created_at']),
        updatedAt: parseNullableDateTime(json['updated_at']),
        cancelledBy: parseNullableInt(json['cancelled_by']),
        cancelledRole: json['cancelled_role'],
      );
    } catch (e) {
      rethrow;
    }
  }
}
